### PR TITLE
Binary Predicate Test Dispatch

### PR DESCRIPTION
**Pull request by thomcom**
## Description
Currently `contains_properly`, `intersection`, and `equals` based predicates are all tested in their own files and lack explicit corner cases and boundary conditions for certain tests. They also do not test unimplemented cases at all.

This PR creates `binpred_test_dispatch.py` and `test_binpred_test_dispatch.py`. These files are used to exhaustively test the space of binary predicates and dispatched feature pairs. This will allow easy implementation of the remaining binary predicates in a test-driven fashion. Finally, failing tests are automatically skipped for now, until the entire suite has been finished. At that point the skip functionality will be removed and fails will be treated as such.

Each test of `feature1.predicate(feature2)` is also dispatched internally to a group of custom values for each feature type. These specified values will be used in the follow up PR for https://github.com/rapidsai/cuspatial/issues/1036, enabling truly exhaustive testing of all possible combinations of features and feature coordinate arrangements.

When this test dispatcher and types in #1036 are finished, the pre-existing tests should be able to be deleted.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cuspatial/cuspatial/tests/binpreds/binpred_test_dispatch.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        "triple_center_disjoint",
+        "border_overlap",
+        "interior_overlap",
+    ]
+)
+def test_type(request):
+    return request.param
+
+
+"""Named features for each feature dispatch."""
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/binpred_test_dispatch.py:98]: Can you give a bit more detail on what basis did you choose these test examples from? I expect what you chose should be sufficient to cover most of  the corner cases for various geometry combination and binpreds.

File: python/cuspatial/cuspatial/tests/binpreds/binpred_test_dispatch.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+points_dispatch = {
+    "feature1": Point(0.0, 0.0),
+    "feature2": Point(1.0, 1.0),
+    "feature1_bo": Point(1.0, 1.0),
+    "feature2_bo": Point(0.0, 0.0),
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/binpred_test_dispatch.py:103]: There are duplicate values in the dict, can you explain why?

File: python/cuspatial/cuspatial/tests/binpreds/binpred_test_dispatch.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+"""The following fixtures are used to generate tests for
+each possible combination of geometry types and binary
+predicates. The fixtures are combined in the test function
+in `test_binpreds_test_dispatch.py` to make the following
+tuple: (predicate, geotype, geotype, expected_result). The
+geotype fixtures are used to generate the first two
+elements of the tuple. The predicate fixture is used to
+generate the third element of the tuple. The expected_result
+fixture is used to generate the fourth element of the tuple.
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/binpred_test_dispatch.py:27]: The actual function signature is a different, it has only 3 arguments.

```suggestion
tuple: (predicate, geotype_tuple, expected_result). The
geotype fixtures are used to generate the second 
elements of the tuple. The predicate fixture is used to
generate the first element of the tuple. The expected_result
fixture is used to generate the third element of the tuple.
```

File: python/cuspatial/cuspatial/tests/binpreds/test_binpred_test_dispatch.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+by the test function. Will be removed when all tests are passing."""
+
+
+def skip_on_exception(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception as e:
+            pytest.skip(f"Skipping due to an exception: {e}")
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_binpred_test_dispatch.py:24]: `xfail` is probably more suitable here. If they are eventually meant to pass.

File: python/cuspatial/cuspatial/tests/binpreds/test_binpred_test_dispatch.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        pd.testing.assert_series_equal(expected, got.to_pandas())
+    except AssertionError as e:
+        print("Binary Predicate Test failed")
+        print("----------------------------")
+        print(f"lhs: {lhs}")
+        print(f"rhs: {rhs}")
+        print(f"predicate: {predicate}")
+        print(f"test_type: {test_type}")
+        print(f"expected: {expected}")
+        print(f"got: {got}")
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_binpred_test_dispatch.py:56]: I was concerned if these prints will be recorded in CI. But probably not since pytest by default doesn't capture stdouts.

File: python/cuspatial/cuspatial/tests/binpreds/test_binpred_test_dispatch.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+
+from functools import wraps
+
+import pandas as pd
+import pytest
+from binpred_test_dispatch import (  # noqa: F401
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_binpred_test_dispatch.py:7]: Why noqa here?

File: python/cuspatial/cuspatial/tests/binpreds/test_binpred_test_dispatch.py (hunk 5 above)
Comment (isVoid) [reply python/cuspatial/cuspatial/tests/binpreds/test_binpred_test_dispatch.py:56]: Even still, I'm hoping perhaps we should disable them during CI. Since they are still run and brings burden.
